library/pkcs7_write.c

```diff
@@ -0,0 +1,843 @@
+              strlen( MBEDTLS_OID_PKCS1_RSA ), 0 );
+    if( ret == 0 ) {
+        ret = pkcs7_write_data(
+                  start, size, ptr, WRITE_OID, (void *) pkcs7_info->hash_funct_oid,
+                  strlen( pkcs7_info->hash_funct_oid ), 0 );
+    }
+
+    return ( ret );
+}
+
+static int pkcs7_set_signer_cert_data( unsigned char **start, size_t *size,
+                                       unsigned char **ptr, mbedtls_pkcs7_info *pkcs7_info,
+                                       mbedtls_x509_crt *pub, unsigned char *priv,
+                                       size_t priv_size )
+{
+    int ret, signed_info_version = 1;
+    size_t bytes_written_in_step, curr_len;
+    unsigned char tag = MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE;
+
+    ret = pkcs7_set_algorithm_ids( start, size, ptr, pkcs7_info, pub, priv,
+                                   priv_size );
+    if( ret != 0 )
+        return ( ret );
+
+    /* add serial */
+    curr_len = *size - ( *ptr - *start );
+    ret = pkcs7_write_data( start, size, ptr,
+                            WRITE_SERIAL,
+                            pub->serial.p, pub->serial.len, 0 );
+    if( ret != 0 ) {
+        mbedtls_printf(
+            "ERROR: Failed to add certificate serial number to Signer info of PKCS7\n" );
+
+        return ( ret );
+    }
+    ret = pkcs7_write_data( start, size, ptr, WRITE_RAW_BUFFER,
+                            pub->issuer_raw.p, pub->issuer_raw.len, 0 );
+    if( ret != 0 ) {
+        mbedtls_printf( "ERROR: Failed to add issuer data to Signer info of PKCS7\n" );
+
+        return ( ret );
+    }
+    /* add info */
+    bytes_written_in_step = *size - ( *ptr - *start ) - curr_len;
+    ret = pkcs7_write_data( start, size, ptr,
+                            WRITE_TAG, &tag,
+                            bytes_written_in_step, 0 );
+    if( ret == 0 ) {
+        /* add signed info version, see https:// tools.ietf.org/html/rfc2315
+         * section 9.2 */
+        ret = pkcs7_write_data( start, size, ptr, WRITE_VERSION,
+                                &signed_info_version, sizeof( signed_info_version ), 0 );
+    }
+    if( ret )
+        mbedtls_printf( "ERROR: Failed to add owner flag to Signer Info of PKCS7\n" );
+
+    return ( ret );
+}
+
+static int pkcs7_set_each_signer_data( unsigned char **start, size_t *size,
+                                       unsigned char **ptr,
+                                       mbedtls_pkcs7_info *pkcs7_info )
+{
+    int ret;
+    size_t bytes_written_in_step, curr_len;
+    unsigned char tag = MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE;
+    mbedtls_x509_crt *x509 = NULL;
+    /* if no signers than quit */
+    if( pkcs7_info->key_pairs < 1 ) {
+        ret = MBEDTLS_ERR_PKCS7_INVALID_SIGNER_INFO;
+        mbedtls_printf( "ERROR: No keys given to sign with\n" );
+    }
+
+    for( int i = 0; i < pkcs7_info->key_pairs; i++ ) {
+        x509 = mbedtls_calloc( 1, sizeof( *x509 ) );
```

Comment:
I believe type of pointer is known at compile time.

Reply:
`sizeof x` doesn't evaluate `x` and doesn't need `x` to have a valid value. It only depends on the type. `sizeof(*p)` is preferable to `sizeof(type)` because it will remain correct if the program is modified and the variable's type changes.